[[ Bug 22918 ]] Fix crash when deleting object target of messages

livecodeian opened the pull request:
This patch fixes a crash that can occur when an object is deleted while
handling a message sent to that object.

Fixes https://quality.livecode.com/show_bug.cgi?id=22918

livecodeian committed:
[[ Bug 22918]] Fix crash when deleting object target of messages

This patch fixes a crash that can occur when an object is deleted while
handling a message sent to that object.

runrevmark commented:
@livecodeian Can you file a new bug for this and update this to reference it. I don't suppose you can see a simple testcase / recipe stack for it too?

runrevmark commented on engine/src/object.cpp:
```diff
@@ -5784,7 +5787,32 @@ void MCDeletedObjectsOnObjectResumeDeletion(MCObject *p_object, void *p_deletion
+	{
+		// If the current pool is nil then the object must have been deleted, so
+		//    remove the reference to the saved deleted object pool and clean up.
+		t_pool->references -= 1;
```
It looks like we do this more than once - perhaps refactor into `MCDeletedObjectPoolRelease`?

livecodeian committed:
[[ Bug 22918 ]] Add release note

This patch adds a release note for bug 22918, where deleting an object
while handling a message sent to that object could cause a crash.

livecodeian commented:
@runrevmark Changes made, bug report filed, and release note added :)

runrevmark commented on engine/src/object.cpp:
```diff
@@ -5719,21 +5755,8 @@ void MCDeletedObjectsOnObjectDeleted(MCObject *p_object)
-            MCDeletedObjectPoolDestroy(t_this_pool);
-        }
-    }
+	t_pool = MCDeletedObjectPoolGetNonDefunct(MCDeletedObjectPoolRelease(t_pool));
```
Rather than making Release return a value, and then having another method wrapping it, it would be clearer to have:
```
/* Release the given object pool, and return the nearest non-defunct ancestor.
 * Any intermediate, unreferenced, defunct pools are destroyed. */
static MCDeletedObjectPool *MCDeletedObjectPoolUnwind(MCDeletedObjectPool *p_pool);
```

runrevmark replied:
Try and make indent with spaces to match the handler.

runrevmark commented on engine/src/object.cpp:
```diff
@@ -5571,6 +5571,46 @@ static void MCDeletedObjectPoolDestroy(MCDeletedObjectPool *p_pool)
+	return p_pool;
+}
+
+static MCDeletedObjectPool* MCDeletedObjectPoolRelease(MCDeletedObjectPool *p_pool)
```
I suggest making this not return a value as it is only used in (top-level) contexts where it is not required (and where it is, we actually want a different pool).

runrevmark replied:
```
/* Release the given pool. If the pool becomes unreferenced as a result,
 * it and any defunct unreferenced ancestors are destroyed. */
```

runrevmark commented on engine/src/object.cpp:
```diff
@@ -5571,6 +5571,46 @@ static void MCDeletedObjectPoolDestroy(MCDeletedObjectPool *p_pool)
+	return MCDeletedObjectPoolCleanup(p_pool);
+}
+
+static MCDeletedObjectPool* MCDeletedObjectPoolGetNonDefunct(MCDeletedObjectPool *p_pool)
```
Change this to Unwind as suggested below, and merge with the body of Release (rather than calling Release a sub-routine - the main work is done in Cleanup which is fine).

runrevmark commented on engine/src/object.cpp:
```diff
@@ -5571,6 +5571,46 @@ static void MCDeletedObjectPoolDestroy(MCDeletedObjectPool *p_pool)
+	return MCDeletedObjectPoolCleanup(p_pool);
+}
+
+static MCDeletedObjectPool* MCDeletedObjectPoolMakeDefunct(MCDeletedObjectPool* p_pool)
```
Don't return a value unless it is used.

runrevmark replied:
```
/* Mark the given pool as defunct, destroying it (and any defunct, unreferenced
ancestors) if it is unreferenced. */
```

runrevmark commented on engine/src/object.cpp:
```diff
@@ -5680,11 +5720,7 @@ void MCDeletedObjectsLeaveWait(bool p_dispatching)
-    // If the objectpool has no references then we can delete it.
-    if (t_pool -> references == 0)
-        MCDeletedObjectPoolDestroy(t_pool);
+	MCDeletedObjectPoolMakeDefunct(t_pool);
```
If you could make this indent with spaces (since the rest of the handler is) that would be better.

runrevmark commented on engine/src/object.cpp:
```diff
@@ -5750,15 +5773,7 @@ void MCDeletedObjectsOnObjectDestroyed(MCObject *p_object)
-        t_pool = t_pool -> parent;
-        MCDeletedObjectPoolDestroy(t_this_pool);
-    }
+	MCDeletedObjectPoolRelease(t_pool);
```
```
/* Release the pool and clean up the ancestor chain where possible. */
```
Also try and make indent with spaces.

runrevmark reviewed (changes requested):
So the refactoring is good - but it has made things a little unclear. Main change is just to tweak the refactored operations to ensure they align strictly with use (rather than being more general then they need to be).

runrevmark commented:
@livecode-vulcan review ok 1353807

livecode-vulcan commented:
💙 review by @runrevmark ok 135380751d4c3ee1e43b157aea68e5f11119407c

livecode-vulcan commented:
😎 test success 135380751d4c3ee1e43b157aea68e5f11119407c

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/871)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/280)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/280)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/280)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/280)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2397)
* try-community-universal-ios-iphoneos14.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.0/builds/0)
* try-community-universal-ios-iphoneos13.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.5/builds/49)
* try-community-universal-ios-iphonesimulator14.0: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.0/builds/0)
* try-community-universal-ios-iphonesimulator13.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.5/builds/47)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2425)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2413)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2412)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2447)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2387)
